gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/activity/impl/DeleteWorkDirsActivityImpl.java

```diff
@@ -0,0 +1,224 @@
+    if (!state.contains(ConfigurationKeys.WRITER_STAGING_DIR) || !state.contains(ConfigurationKeys.WRITER_OUTPUT_DIR)) {
+      return Maps.newHashMap();
+    }
+    String writerFsUri = state.getProp(ConfigurationKeys.WRITER_FILE_SYSTEM_URI, ConfigurationKeys.LOCAL_FS_URI);
+    FileSystem fs = JobLauncherUtils.getFsWithProxy(state, writerFsUri, WriterUtils.getFsConfiguration(state));
+    Map<String, Long> cleanupMetrics = Maps.newHashMap();
+    String jobId = state.getJobId();
+
+    for (String resource : resourcesToClean) {
+      Path pathToClean = new Path(resource);
+      Long filesCleaned = fs.getContentSummary(pathToClean).getLength();
+      log.info("Cleaning up resource directory " + pathToClean);
+      if (!isDirectoryJobLevelScoped(resource, jobId)) {
+        HadoopUtils.deletePath(fs, pathToClean, true);
+        cleanupMetrics.put(pathToClean.toString(), filesCleaned);
+      } else {
+        log.warn("Not deleting directory " + pathToClean.toUri().getPath() + " as it is not job level scoped");
+      }
+    }
+
+    Set<Path> singleLevelParentDirs = resourcesToClean.stream().map(resource -> new Path(resource).getParent()).collect(
+        Collectors.toSet());
+
+    for (Path parentDir : singleLevelParentDirs) {
+      if (fs.exists(parentDir) && fs.listStatus(parentDir).length == 0) {
+        log.info("Deleting directory " + parentDir);
+        if (!isDirectoryJobLevelScoped(parentDir.toString(), jobId)) {
+          HadoopUtils.deletePath(fs, parentDir, true);
+          cleanupMetrics.put(parentDir.toString(), 0L);
+        } else {
+          log.warn("Not deleting directory " + parentDir.toUri().getPath() + " as it is not job level scoped");
+        }
+      }
+    }
+    return cleanupMetrics;
+  }
+
+  public static Map<String, Long> cleanTaskStagingData(TaskState state, Logger log, Closer closer,
+      Map<String, ParallelRunner> parallelRunners) throws IOException {
+    int numBranches = state.getPropAsInt(ConfigurationKeys.FORK_BRANCHES_KEY, 1);
+
+    int parallelRunnerThreads =
+        state.getPropAsInt(ParallelRunner.PARALLEL_RUNNER_THREADS_KEY, ParallelRunner.DEFAULT_PARALLEL_RUNNER_THREADS);
+
+    Map<String, Long> cleanupMetrics = Maps.newHashMap();
+    String jobId = state.getJobId();
+    for (int branchId = 0; branchId < numBranches; branchId++) {
+      String writerFsUri = state.getProp(
+          ForkOperatorUtils.getPropertyNameForBranch(ConfigurationKeys.WRITER_FILE_SYSTEM_URI, numBranches, branchId),
+          ConfigurationKeys.LOCAL_FS_URI);
+      FileSystem fs = JobLauncherUtils.getFsWithProxy(state, writerFsUri, WriterUtils.getFsConfiguration(state));
+      ParallelRunner parallelRunner = JobLauncherUtils.getParallelRunner(fs, closer, parallelRunnerThreads, parallelRunners);
+
+      Path stagingPath = WriterUtils.getWriterStagingDir(state, numBranches, branchId);
```

Comment:
what's different about this vs. processing each `WorkUnit` to pull out the staging dir location?  when the `GenerateWUs`  activity did that for all of the WUs and put them into a list, how would that differ from doing them one-by-one here?